### PR TITLE
[RealImage/apsara #740] Plugin`s basic implementation

### PR DESCRIPTION
**Pull request by VValkovoy**
# What this PR does?
* Creates new `gst-plugin-imbsink` Gstreamer plugin.

- **Issue**
1) Current functionality of streaming media buffers to IMB is done inside ImbSinkBin that should be moved to a separate plugin.
2) Synchronous feeding IMB with media buffers should be simplified and plugin`s architecture should be designed accordingly. 

- **Solution**
    * Move connection with IMB and feeding it with media buffers from ImbSinkBin to a new separate plugin `gst-plugin-imbsink`.
    * Move `mvc201toolsapi` to `gst-plugin-imbsink` Gstreamer sink plugin.
    * Remove internal dependencies and functions which are redundant by their functionality.
    * Enable required playback functionality with basic navigation as Play/Pause/Stop.

- **Expected behavior**
With new architecture Plugin ImbSink should keep the only functionality - connection with IMB and streaming input  video and audio buffers for decoding aboard IMB.

# How to build:
* `git clone --recursive -b feature/stream-imb git@github.com:RealImage/gst-plugin-imbsink.git`
* `cmake  ..`
* `make -j4`

### CODE REVIEW

File: src/IStreamSource.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    unsigned int PadSize; // Needed when encrypted only
+
+    Buffer()
+        : Data(NULL)
+        , Size(0)
+        , EncryptMode(Qube::EncryptionMode::Unknown)
+        , Index(0)
+        , PlainTextOffset(0)
+        , SourceLength(0)
+        , PadSize(0)
```
Comment (AntonQubecinema) [inline src/IStreamSource.h]: Init all values

File: src/IStreamSource.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    {
+    }
+
+    virtual ~Buffer() {}
```
Comment (AntonQubecinema) [inline src/IStreamSource.h]:
```suggestion
    virtual ~Buffer() = default;
```

File: src/IStreamSource.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
+/*
+ * Virtual class for Stream Source of Gstreamer plugins
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
```
Comment (AntonQubecinema) [inline src/IStreamSource.h]: consider renaming

File: src/Media.h [hunk 4]
```diff
@@ -0,0 +1,216 @@
+/*
+ * Full Media information
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
+
+#ifndef __QUBE_MEDIA_H__
```
Comment (AntonQubecinema) [inline src/Media.h]:
```suggestion
#ifndef MEDIA_H
```

File: src/MediaTypes.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
+#ifndef MEDIA_TYPES_H
+#define MEDIA_TYPES_H
+
+#include <string>
```
Comment (AntonQubecinema) [inline src/MediaTypes.h]: Unused

File: src/Mvc20xMediaBlock.cpp [hunk 6]
```diff
@@ -0,0 +1,2434 @@
...
+#include <boost/scoped_ptr.hpp>
+#include <boost/uuid/string_generator.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/uuid/nil_generator.hpp>
+
+#include <gst/gst.h>
+#include <queue>
+
+#define byte uint8_t
```
Comment (AntonQubecinema) [inline src/Mvc20xMediaBlock.cpp]: `using`

File: src/Mvc20xMediaBlock.cpp [hunk 7]
```diff
@@ -0,0 +1,2434 @@
...
+                         message.append(FormattedString(format, ##__VA_ARGS__));                                       \
+                         message.append(FormattedString(" : (%d) ", result));                                          \
+                         const char *errString = mvc2::MvcDevice::getErrorDescription(result);                         \
+                         if(errString)                                                                                 \
+                             message.append(errString);                                                                \
+                         GST_ERROR("%s", message.c_str());                                                             \
+                         throw std::exception();                                                                       \
+                      }
+
+typedef unsigned long DWORD;
```
Comment (AntonQubecinema) [inline src/Mvc20xMediaBlock.cpp]: `using`

File: src/ProductModel.h [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Product Model information
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
+
+#ifndef __QUBE_PRODUCT_TYPE_H__
```
Comment (AntonQubecinema) [inline src/ProductModel.h]:
```suggestion
#ifndef PRODUCT_TYPE_H
```

File: src/ProductModel.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Product Model information
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
+
+#ifndef __QUBE_PRODUCT_TYPE_H__
+#define __QUBE_PRODUCT_TYPE_H__
+
+enum ProductModel
```
Comment (AntonQubecinema) [inline src/ProductModel.h]: Do we need this?

File: src/StringUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+void StringUtils::FormatString(std::string& targetBuffer, const char* format, va_list vargs)
+{
+    // Allocate a buffer on the stack that's big enough for us almost
+    // all the time.  Be prepared to allocate dynamically if it doesn't fit.
+    size_t            size = 1024;
+    char              stackbuf[1024];
+    std::vector<char> dynamicbuf;
+    char*             buf = &stackbuf[0];
+
+    for (;;)
```
Comment (AntonQubecinema) [inline src/StringUtils.cpp]: Infinite loop is a bad solution

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+if(NOT MSVC)
+    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DDEBUG -ggdb")
+    if (!APPLE)
+        set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -static-libasan")
+    else(!APPLE)
+        add_compile_options(-Wno-delete-non-virtual-dtor)
+    endif(!APPLE)
+endif()
+
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply CMakeLists.txt]: Removed

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+if (NOT PLUGIN_INSTALL_DIR)
+    message(STATUS "No plugin install dir specified, default to 'lib/${ARCH}-linux-gnu/gstreamer-1.0/'")
+    set(PLUGIN_INSTALL_DIR "lib/${ARCH}-linux-gnu/gstreamer-1.0/")
+endif()
+
+# dependencies
+find_package(GLIB REQUIRED)
+find_package(GStreamer 1.8 COMPONENTS base REQUIRED)
+find_package(OpenSSL REQUIRED)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply CMakeLists.txt]: Set EXACT version.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE "amd64")
+endif()
+
+if (NOT PLUGIN_INSTALL_DIR)
+    message(STATUS "No plugin install dir specified, default to 'lib/${ARCH}-linux-gnu/gstreamer-1.0/'")
+    set(PLUGIN_INSTALL_DIR "lib/${ARCH}-linux-gnu/gstreamer-1.0/")
+endif()
+
+# dependencies
+find_package(GLIB REQUIRED)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply CMakeLists.txt]: Set version.

File: CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+find_package(GStreamer 1.8 COMPONENTS base REQUIRED)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply CMakeLists.txt]: Yes.

File: README.md [hunk 15]
```diff
@@ -1,3 +1,185 @@
...
+  SrcMode             : Sets the sample rate converter's output sample rate.
+                        flags: readable, writable
+                        Integer. Range: 0 - 2147483647 Default: 48000 
+  SignalFormat        : Sets the video output signal format.
+                        flags: readable, writable
+                        Integer. Range: 0 - 2147483647 Default: 0 
+  StereoPlayback      : Sets whether playback is stereo/3D.
+                        flags: readable, writable
+                        Integer. Range: 0 - 2147483647 Default: 0 
+  Xi                  : Sets one or more Xi IP address(es).
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply README.md]: Renamed to Xi-url.

File: README.md [hunk 16]
```diff
@@ -1,3 +1,185 @@
-# gst-plugin-imbsink
+# Overview
+`imbsink` is gstreamer sink plugin which sends frames over to IMB
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply README.md]: Removed.

File: cmake/Modules/FindMVC201ToolsApi.cmake [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  URL "${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software"
+  CONFIGURE_COMMAND ${MVC201_TOOLS_CONFIGURE_COMMAND} COMMAND ${MVC201_TOOLS_CONFIGURE_COMMAND}
+  INSTALL_COMMAND ""
+  TEST_COMMAND ""
+)
+
+# Export
+set(MVC201_TOOLS_LIBRARY ${DEPENDENCIES_DIR}/src/mvc201-tools-api-build/bin/libmvc2api.so)
+set(MVC201_TOOLS_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/service/mvc2api/include
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/include
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply cmake/Modules/FindMVC201ToolsApi.cmake]: Used relative path for mvc libs. Removed redundant dir inludes.

File: cmake/Modules/FindMVC201ToolsApi.cmake [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+ExternalProject_Add(mvc201-tools-api
+  PREFIX ${DEPENDENCIES_DIR}
+  URL "${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software"
+  CONFIGURE_COMMAND ${MVC201_TOOLS_CONFIGURE_COMMAND} COMMAND ${MVC201_TOOLS_CONFIGURE_COMMAND}
+  INSTALL_COMMAND ""
+  TEST_COMMAND ""
+)
+
+# Export
+set(MVC201_TOOLS_LIBRARY ${DEPENDENCIES_DIR}/src/mvc201-tools-api-build/bin/libmvc2api.so)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply cmake/Modules/FindMVC201ToolsApi.cmake]: Added lib itself.

File: cmake/Modules/FindMVC201ToolsApi.cmake [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/common/mm_stdlib/include/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/common/FilenameCounter/include/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/common/
+#                             ${PROJECT_SOURCE_DIR}/deps/mvc201toolsapi/current/software/application/mvc2ipserver/src/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/application/mvc2play/src/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/3rd_party/ixml/include/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/service/mvc2playback/include/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/application/mvc2upnprenderer/src/
+                             ${PROJECT_SOURCE_DIR}/deps/mvc201-tools-api/current/software/modules/common/mpeg_toolbox/include/)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply cmake/Modules/FindMVC201ToolsApi.cmake]: Added

File: src/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+include_directories(SYSTEM ${GLIB_INCLUDE_DIRS}
+                           ${GLIBCONFIG_INCLUDE_DIR}
+                           ${GSTREAMER_INCLUDE_DIRS}
+                           ${GSTREAMER_BASE_INCLUDE_DIRS}
+                           ${OPENSSL_INCLUDE_DIR}
+                           ${Boost_INCLUDE_DIRS}
+                           ${MVC201_TOOLS_INCLUDE_DIR})
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR}
+                    ${MVC201_TOOLS_INCLUDE_DIR})
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/CMakeLists.txt]: Removed MVC include dir to system.

File: src/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+
+set(IMB_SINK_TARGET_NAME gstimbsink)
+add_library(${IMB_SINK_TARGET_NAME}
+            ${PLUGIN_BUILD_TYPE}
+            ${IMB_SINK_SRC_LIST}
+            ${IMB_SINK_INC_LIST})
+
+target_link_libraries(${IMB_SINK_TARGET_NAME} ${COMMON_LIBRARIES}
+                      ${CMAKE_THREAD_LIBS_INIT}
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/CMakeLists.txt]: Removed linking thread libs.

File: src/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+                      ${OPENSSL_LIBRARIES})
+
+target_link_libraries(${IMB_SINK_TARGET_NAME}
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/CMakeLists.txt]: Combined.

File: src/Condition.h [hunk 23]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Boost threading management
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
+
+#ifndef __QUBE_CONDITION_H__
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/Condition.h]: Changed

File: src/GstBufferMap.h [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+{
+public:
+    // maps the one or more memory of the given GstBuffer and unmaps it when the object of GstBufferMap is
+    // destroyed the startMemoryIndex and memoryCount allows us to control the range of memories to be mapped.
+    // when memoryCount is -1 then all the remaining memories from/including startMemoryIndex are mapped
+    GstBufferMap(GstBuffer* buf, GstMapFlags flag, int startMemoryIndex = 0, int memoryCount = -1);
+    ~GstBufferMap();
+
+    guint8* GetData();
+    gsize   GetDataSize();
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstBufferMap.h]: Added

File: src/GstBufferMap.h [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+    gsize   GetBufferSize();
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstBufferMap.h]: Added

File: src/GstImbSink.cpp [hunk 26]
```diff
@@ -0,0 +1,896 @@
...
+ */
+static gboolean initialize_media_block(ImbSink* self)
+{
+    self->is_media_block_configured = false;
+
+    if (self->media_block == nullptr)
+    {
+        try
+        {
+            if (IsXpI(self->product_model))
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Will be removed with the next PR on code refactor

File: src/GstImbSink.cpp [hunk 27]
```diff
@@ -0,0 +1,896 @@
...
+
+static gboolean      imb_sink_start(GstBaseSink* sink);
+static GstFlowReturn imb_sink_render(GstBaseSink* sink, GstBuffer* buffer) noexcept;
+
+static GstStateChangeReturn imb_sink_change_state_func(GstElement* element, GstStateChange transition);
+static gboolean             create_configure_renderers(ImbSink* self);
+static gboolean             configure_video_renderer(ImbSink* self, GstBuffer* gstVideoBuffer);
+static gboolean             configure_audio_renderer(ImbSink* self, GstBuffer* gstAudioBuffer);
+
+static gboolean play_renderers(ImbSink* self /*, GstBuffer* gstVideoBuffer, GstBuffer* gstAudioBuffer*/);
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 28]
```diff
@@ -0,0 +1,896 @@
...
+{
+    self->is_media_block_configured = false;
+
+    if (self->media_block == nullptr)
+    {
+        try
+        {
+            if (IsXpI(self->product_model))
+            {
+                self->media_block = new MediaBlock::Mvc20x();
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Due to internally thrown exceptions.

File: src/GstImbSink.cpp [hunk 29]
```diff
@@ -0,0 +1,896 @@
...
+}
+
+static void imb_sink_set_property(GObject* object, guint propId, const GValue* value, GParamSpec* pspec)
+{
+    ImbSink* self = IMB_SINK(object);
+
+    switch (propId)
+    {
+
+        case PROP_OFFSET:
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed.

File: src/GstImbSink.cpp [hunk 30]
```diff
@@ -0,0 +1,896 @@
...
+            self->audio_delay = g_value_get_int(value);
+            GST_INFO_OBJECT(self, "Audio Delay set to: %d", self->audio_delay);
+            break;
+        }
+
+        case PROP_SRC_MODE:
+        {
+            int src_mode = g_value_get_int(value);
+
+            if (src_mode == 0)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Used switch.

File: src/GstImbSink.cpp [hunk 31]
```diff
@@ -0,0 +1,896 @@
...
+    {
+        case PROP_OFFSET:
+        {
+            g_value_set_double(value, self->offset);
+            break;
+        }
+
+        case PROP_CURRENT_POS:
+        {
+            if (self->media_block && (self->is_media_block_configured == true))
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 32]
```diff
@@ -0,0 +1,896 @@
...
+            {
+                self->cur_pos = self->media_block->GetPosition();
+            }
+            else
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed.

File: src/GstImbSink.cpp [hunk 33]
```diff
@@ -0,0 +1,896 @@
...
+
+        case PROP_SIGNAL_FORMAT:
+        {
+            g_value_set_int(value, (int)self->signal_format);
+            break;
+        }
+
+        case PROP_STEREO_PLAYBACK:
+        {
+            g_value_set_int(value, self->is_stereo_playback ? TRUE : FALSE);
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed.

File: src/GstImbSink.cpp [hunk 34]
```diff
@@ -0,0 +1,896 @@
...
+            break;
+    }
+}
+
+static gboolean imb_sink_start(GstBaseSink* sink)
+{
+    ImbSink* self = IMB_SINK(sink);
+
+    /* Initialize Video and Audio params with default values */
+    if (self->media_block == nullptr)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 35]
```diff
@@ -0,0 +1,896 @@
...
+    {
+        if (!initialize_media_block(self))
+            return FALSE;
+    }
+
+    return TRUE;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Modified

File: src/GstImbSink.cpp [hunk 36]
```diff
@@ -0,0 +1,896 @@
...
+}
+
+static gboolean imb_sink_start(GstBaseSink* sink)
+{
+    ImbSink* self = IMB_SINK(sink);
+
+    /* Initialize Video and Audio params with default values */
+    if (self->media_block == nullptr)
+    {
+        if (!initialize_media_block(self))
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 37]
```diff
@@ -0,0 +1,896 @@
...
+}
+
+static GstFlowReturn imb_sink_render(GstBaseSink* sink, GstBuffer* buffer) noexcept
+{
+    ImbSink* self = IMB_SINK(sink);
+
+    GST_INFO("ImbSink rendering input buffer PTS % " GST_TIME_FORMAT " Size [%lu]",
+             GST_TIME_ARGS(GST_BUFFER_DTS(buffer)), gst_buffer_get_size(buffer));
+
+    if (self->is_media_block_playing == false)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: No. It is needed for Play/Pause/Stop playback control.

File: src/GstImbSink.cpp [hunk 38]
```diff
@@ -0,0 +1,896 @@
...
+        {
+            return GST_FLOW_ERROR;
+        }
+    }
+
+    /* Define media type from buffer`s Meta data */
+    const QubeMetaInfo* qube_meta_info = GST_QUBE_META_GET(buffer);
+    if (qube_meta_info == nullptr)
+    {
+        GST_ERROR_OBJECT(self, "Unable1 to get meta from the buffer");
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Changed

File: src/GstImbSink.cpp [hunk 39]
```diff
@@ -0,0 +1,896 @@
...
+{
+    self->cur_pos = 0;
+    self->caps = nullptr;
+
+    self->audio_delay = 0;
+    self->offset = 0;
+    self->sample_rate_converter_mode = static_cast<Qube::Audio::SamplingRate::Enum>(48000);
+    self->signal_format = static_cast<Qube::Video::SignalFormat::Type>(0);
+    self->is_stereo_playback = false;
+    self->xi = nullptr;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Added freeing properties in finalize().

File: src/GstImbSink.cpp [hunk 40]
```diff
@@ -0,0 +1,896 @@
...
+    /*
+     * debug category for filtering log messages
+     */
+    GST_DEBUG_CATEGORY_INIT(imb_sink_debug, IMB_SINK_NAME, 0, IMB_SINK_DESC);
+    return gst_element_register(videoSink, IMB_SINK_NAME, GST_RANK_NONE, IMB_SINK_TYPE);
+}
+
+static gboolean create_configure_renderers(ImbSink* self)
+{
+    if (self->media_block || (self->is_media_block_configured == false))
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Changed.

File: src/GstImbSink.cpp [hunk 41]
```diff
@@ -0,0 +1,896 @@
...
+        {
+            GST_ERROR_OBJECT(self, "Configuring media block failed: %s", e.what());
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static gboolean play_renderers(ImbSink* self /*, GstBuffer* gstVideoBuffer, GstBuffer* gstAudioBuffer*/)
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 42]
```diff
@@ -0,0 +1,896 @@
...
+        GST_ERROR_OBJECT(self, "Media block play failed: %s", e.what());
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+void destroy_renderers(ImbSink* self)
+{
+    try
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Mvc20x::Stop(bool flush) and Mvc20x::Stop() are different functions.

File: src/GstImbSink.cpp [hunk 43]
```diff
@@ -0,0 +1,896 @@
...
+    switch (transition)
+    {
+        case GST_STATE_CHANGE_READY_TO_PAUSED:
+        {
+            GST_INFO_OBJECT(self, "...............QSB: READY TO PAUSED ...............");
+
+            if (!self->is_media_block_configured)
+            {
+                create_configure_renderers(self);
+
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Removed

File: src/GstImbSink.cpp [hunk 44]
```diff
@@ -0,0 +1,896 @@
...
+            self->is_media_block_playing = false;
+            self->is_media_block_configured = false;
+
+            GST_INFO_OBJECT(self, "Done cleaning up resources.");
+
+            break;
+        }
+
+        default:
+            break;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: No. We have other state change transitions which we do not handle and let Gstreamer manage them by itself.

File: src/GstImbSink.cpp [hunk 45]
```diff
@@ -0,0 +1,896 @@
...
+        }
+
+        case GST_STATE_CHANGE_PLAYING_TO_PAUSED:
+        {
+            GST_INFO_OBJECT(self, "...............QSB: PLAYING TO PAUSED ...............");
+
+            pause_renderers(self);
+
+            self->is_media_block_playing = false;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Left as it is inside _state_func().

File: src/GstImbSink.cpp [hunk 46]
```diff
@@ -0,0 +1,896 @@
...
+            break;
+        }
+
+        case GST_STATE_CHANGE_PAUSED_TO_READY:
+        {
+            GST_INFO_OBJECT(self, "...............QSB: PAUSED TO READY ...............");
+
+            stop_renderers(self);
+
+            self->is_media_block_playing = false;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Left as it is inside _state_func().

File: src/IMediaBlock.h [hunk 47]
```diff
@@ -0,0 +1,103 @@
...
+    gboolean EnableAuro;
+    guint64  MinTxSize;
+    guint64  IcpPortNumber;
+    gboolean DisableSmpteChannelAssignments;
+    guint64  SeamfullBlankInterval;
+};
+
+struct MediaConfig
+{
+    Qube::Video::Config Video;
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/IMediaBlock.h]: Video and audio configs are totally different.
https://github.com/RealImage/gst-plugin-imbsink/pull/3/files#diff-c37ba6f36ca51e878ee1073da8111a48R55

File: src/IMediaBlock.h [hunk 48]
```diff
@@ -0,0 +1,103 @@
...
+    Qube::Audio::Config Audio;
+};
+
+class IMediaBlock
```
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/IMediaBlock.h]: Will be removed with the next PR upon code refactor.

File: src/IStreamSource.h (hunk 1 above)
Comment (VValkovoy) [reply src/IStreamSource.h]: Added values initialization.

File: src/IStreamSource.h (hunk 2 above)
Comment (VValkovoy) [reply src/IStreamSource.h]: Changed

File: src/IStreamSource.h (hunk 3 above)
Comment (VValkovoy) [reply src/IStreamSource.h]: Renamed.

File: src/Media.h (hunk 4 above)
Comment (VValkovoy) [reply src/Media.h]: Renamed

File: src/MediaTypes.h (hunk 5 above)
Comment (VValkovoy) [reply src/MediaTypes.h]: Removed

File: src/Mvc20xMediaBlock.cpp (hunk 6 above)
Comment (VValkovoy) [reply src/Mvc20xMediaBlock.cpp]: Will be used in the next PR

File: src/Mvc20xMediaBlock.cpp (hunk 7 above)
Comment (VValkovoy) [reply src/Mvc20xMediaBlock.cpp]: Will be used in the next PR

File: src/ProductModel.h (hunk 8 above)
Comment (VValkovoy) [reply src/ProductModel.h]: Renamed

File: src/ProductModel.h (hunk 9 above)
Comment (VValkovoy) [reply src/ProductModel.h]: Ported code. Will be removed with the next PR.

File: src/StringUtils.cpp (hunk 10 above)
Comment (VValkovoy) [reply src/StringUtils.cpp]: Will be refactored with the next PR.

File: src/GstImbSink.cpp (hunk 34 above)
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Already removed.

File: src/GstImbSink.cpp (hunk 35 above)
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Changed

File: src/Mvc20xMediaBlock.cpp (hunk 6 above)
Comment (VValkovoy) [reply src/Mvc20xMediaBlock.cpp]: Compilation error with using.

File: src/GstImbSink.cpp (hunk 42 above)
[the remark this answers is not in the document]
Comment (AntonQubecinema) [reply src/GstImbSink.cpp]: Nope

File: src/StringUtils.cpp (hunk 10 above)
Comment (AntonQubecinema) [reply src/StringUtils.cpp]: Put TODO

File: src/GstImbSink.cpp (hunk 42 above)
[the remark this answers is not in the document]
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Fixed

File: src/GstImbSink.cpp [hunk 49]
```diff
@@ -0,0 +1,899 @@
...
+    ImbSink* self = IMB_SINK(sink);
+
+    GST_INFO("ImbSink rendering input buffer PTS % " GST_TIME_FORMAT " Size [%lu]",
+             GST_TIME_ARGS(GST_BUFFER_DTS(buffer)), gst_buffer_get_size(buffer));
+
+    if (!self->is_media_block_playing)
+    {
+        self->is_media_block_playing = true;
+
+        if (!play_renderers(self))
```
Comment (Iyer-VivekAnand) [inline src/GstImbSink.cpp]: We can call it when gstreamer state changes to PLAYING state.

File: src/GstImbSink.cpp [hunk 50]
```diff
@@ -0,0 +1,899 @@
...
+    return gst_element_register(videoSink, IMB_SINK_NAME, GST_RANK_NONE, IMB_SINK_TYPE);
+}
+
+static gboolean create_configure_renderers(ImbSink* self)
+{
+    if (self->media_block || !self->is_media_block_configured)
+    {
+        try
+        {
+            GST_INFO_OBJECT(self, "Configuring Media block ...");
```
Comment (Iyer-VivekAnand) [inline src/GstImbSink.cpp]: Change it to debug.

File: src/GstImbSink.cpp [hunk 51]
```diff
@@ -0,0 +1,899 @@
...
+            MediaBlock::Config config(self->audio_delay, self->offset, self->sample_rate_converter_mode,
+                                      self->signal_format, self->is_stereo_playback, self->xi,
+                                      self->three_d_enhancement_mode, self->three_d_enhancement_values, self->tcgd,
+                                      self->ltc_aes_pair, self->master, self->slave, self->multi_single_projector_3d,
+                                      self->enable_auro, self->min_tx_size, self->icp_port_number,
+                                      self->disable_smpte_channel_assignments, self->seamfull_blank_interval);
+
+            self->media_block->Configure(config);
+
+            GST_INFO_OBJECT(self, "Media block configured");
```
Comment (Iyer-VivekAnand) [inline src/GstImbSink.cpp]: DEBUG

File: src/ProductModel.h [hunk 52]
```diff
@@ -0,0 +1,26 @@
...
+ * Copyright (c) 2019 Qube Cinema, Inc.
+ */
+
+#ifndef PRODUCT_TYPE_H
+#define PRODUCT_TYPE_H
+
+enum ProductModel
+{
+    QubeXpI,
+    QubeXpD,
```
Comment (Iyer-VivekAnand) [inline src/ProductModel.h]: Remove ProductModels

File: src/StringUtils.cpp [hunk 53]
```diff
@@ -0,0 +1,70 @@
+#include "StringUtils.h"
+#include <vector>
+
+const char* StringUtils::TRUE_STR = "True";
+const char* StringUtils::FALSE_STR = "False";
+
+void StringUtils::FormatString(std::string& targetBuffer, const char* format, va_list vargs)
```
Comment (Iyer-VivekAnand) [inline src/StringUtils.cpp]: Can we use any alternative for this?

File: src/GstImbSink.cpp (hunk 49 above)
Comment (VValkovoy) [reply src/GstImbSink.cpp]: Will be refactored with the next PR.

File: src/GstImbSink.cpp (hunk 50 above)
Comment (VValkovoy) [reply src/GstImbSink.cpp]: We keep Info level for plugin's messages and Debug mostly for internal (libMvc) call messages.

File: src/GstImbSink.cpp (hunk 51 above)
Comment (VValkovoy) [reply src/GstImbSink.cpp]: As mentioned above.

File: src/ProductModel.h (hunk 52 above)
Comment (VValkovoy) [reply src/ProductModel.h]: Ported legacy option.
Will be removed with the next PR.

File: src/StringUtils.cpp (hunk 53 above)
Comment (VValkovoy) [reply src/StringUtils.cpp]: Will be refactored with the next PR.

**Pull request merged**